File: dispatch/ios/CLAID/Sources/CLAID/RemoteFunction/RemoteFunctionHandler.swift
//
//  RemoteFunctionHandler.swift
//  CLAID
//
//  Created by Patrick Langer on 21.02.2025.
//



public actor RemoteFunctionHandler {
    public var futuresHandler = FutureHandler()
    public var toMiddlewareQueue: AsyncStream<Claidservice_DataPackage>.Continuation!

    init(toMiddlewareQueue: AsyncStream<Claidservice_DataPackage>.Continuation) {
        self.toMiddlewareQueue = toMiddlewareQueue
    }
    
    public func makeRemoteFunctionIdentifier(
        runtime: Claidservice_Runtime,
        functionName: String
    ) -> Claidservice_RemoteFunctionIdentifier {
        var remoteFunctionIdentifier = Claidservice_RemoteFunctionIdentifier()
        remoteFunctionIdentifier.functionName = functionName
        remoteFunctionIdentifier.runtime = runtime
        return remoteFunctionIdentifier
    }
    
    public func makeRemoteFunctionIdentifier(
        moduleId: String,
        functionName: String)
    -> Claidservice_RemoteFunctionIdentifier {
        var remoteFunctionIdentifier = Claidservice_RemoteFunctionIdentifier()
        remoteFunctionIdentifier.functionName = functionName
        remoteFunctionIdentifier.moduleID = moduleId
        return remoteFunctionIdentifier
    }
    
    func mapRuntimeFunction<Return, each Parameter>(
        runtime: Claidservice_Runtime,
        functionName: String,
        returnDataTypeExample: Return,
        parameterTypeExamples: repeat each Parameter
    ) -> RemoteFunction<Return, repeat each Parameter> {
        
        return RemoteFunction<Return, repeat each Parameter>(
            futuresHandler: futuresHandler,
            toMiddlewareQueue: toMiddlewareQueue,
            remoteFunctionIdentifier: makeRemoteFunctionIdentifier(runtime: runtime, functionName: functionName),
            returnTypeExample: returnDataTypeExample,
            parameterTypeExamples: repeat each parameterTypeExamples
        )
    }
    
    func mapModuleFunction<Return, each Parameter>(
        targetModule: String,
        functionName: String,
        returnDataTypeExample: Return,
        parameterTypeExamples: repeat each Parameter
    ) -> RemoteFunction<Return, repeat each Parameter> {

        return RemoteFunction<Return, repeat each Parameter>(
            futuresHandler: futuresHandler,
            toMiddlewareQueue: toMiddlewareQueue,
            remoteFunctionIdentifier: makeRemoteFunctionIdentifier(moduleId: targetModule, functionName: functionName),
            returnTypeExample: returnDataTypeExample,
            parameterTypeExamples: repeat each parameterTypeExamples
        )
    }
    
    func handleResponse(_ remoteFunctionResponse: Claidservice_DataPackage) async {
        guard remoteFunctionResponse.controlVal.hasRemoteFunctionReturn else {
            print("Failed to handle remote function response. Did not find RemoteFunctionReturn data")
            return
        }
        Logger.logInfo("RemoteFunctionHandler handling response 1")

        let remoteFunctionReturn = remoteFunctionResponse.controlVal.remoteFunctionReturn
        let futureIdentifier = remoteFunctionReturn.remoteFutureIdentifier
        let uniqueIdentifier = FutureUniqueIdentifier.fromString(futureIdentifier)
        Logger.logInfo("RemoteFunctionHandler handling response 2")

        guard let future = await futuresHandler.lookupFuture(identifier: uniqueIdentifier) else {
            return
        }
        Logger.logInfo("RemoteFunctionHandler handling response 3")

        if remoteFunctionReturn.executionStatus != .statusOk {
            Logger.logWarning("Remote function failed. Future with identifier \(futureIdentifier) failed with status \(remoteFunctionReturn.executionStatus)")
            await future.setFailed()
            return
        }
        Logger.logInfo("RemoteFunctionHandler handling response 4")

        await future.setResponse(remoteFunctionResponse)
        Logger.logInfo("RemoteFunctionHandler handling response 5")

    }
}

File: dispatch/python/claid/remote_function/remote_function_handler.py
from dispatch.proto.claidservice import *
import asyncio
from remote_function.remote_function import RemoteFunction
from remote_function.future_handler import FutureHandler
from remote_function.future_unique_identifier import FutureUniqueIdentifier

from typing import Any
from logger.logger import Logger

class RemoteFunctionHandler:
    def __init__(self, to_middleware_queue: asyncio.Queue):
        self.futures_handler = FutureHandler()
        self.to_middleware_queue = to_middleware_queue

    def map_runtime_function(
        self, 
        runtime: Runtime, 
        function_name: str, 
        return_data_type_example: Any, 
        *parameter_type_examples: Any
    ) -> RemoteFunction:
        return RemoteFunction(
            self.futures_handler,
            self.to_middleware_queue,
            RemoteFunctionIdentifier(runtime=runtime, function_name=function_name),
            return_data_type_example,
            parameter_type_examples
        )

    def map_module_function(
        self, 
        target_module: str, 
        function_name: str, 
        return_data_type_example: Any, 
        *parameter_type_examples: Any
    ) -> RemoteFunction:
        return RemoteFunction(
            self.futures_handler,
            self.to_middleware_queue,
            RemoteFunctionIdentifier(module_id=target_module, function_name=function_name),
            return_data_type_example,
            parameter_type_examples
        )

    async def handle_response(self, remote_function_response: DataPackage):
        print("response ", remote_function_response)
        if not hasattr(remote_function_response.control_val, 'remote_function_return'):
            Logger.log_fatal("Failed to handle remote function response. Did not find RemoteFunctionReturn data")
            return

        Logger.log_info("RemoteFunctionHandler handling response 1")

        remote_function_return = remote_function_response.control_val.remote_function_return
        future_identifier = remote_function_return.remote_future_identifier
        unique_identifier = FutureUniqueIdentifier.from_string(future_identifier)
        Logger.log_info("RemoteFunctionHandler handling response 2")

        future = self.futures_handler.lookup_future(unique_identifier)
        if future is None:
            Logger.log_error(f"Cannot complete future, future with identifier {unique_identifier} is unknown.")
            return
        Logger.log_info("RemoteFunctionHandler handling response 3")

        if remote_function_return.execution_status != RemoteFunctionStatus.STATUS_OK:
            Logger.log_warning(f"Remote function failed. Future with identifier {future_identifier} failed with status {remote_function_return.execution_status}")
            future.set_failed()
            return
        Logger.log_info("RemoteFunctionHandler handling response 4")

        await future.set_response(remote_function_response)
        Logger.log_info("RemoteFunctionHandler handling response 5")